addons/barcodes/static/src/js/barcode_events.js

```diff
@@ -0,0 +1,148 @@
+"use strict";
+
+var core = require('web.core');
+var mixins = core.mixins;
+
+var BarcodeEvents = core.Class.extend(mixins.PropertiesMixin, {
+    timeout: null,
+    key_pressed: {},
+    buffered_key_events: [],
+    // Regexp to match a barcode input and extract its payload
+    // Note: to build in init() if prefix/suffix can be configured
+    regexp: /(.{3,})[\n\r\t]+/,
+    // By knowing the terminal character we can interpret buffered keys
+    // as a barcode as soon as it's encountered (instead of waiting x ms)
+    suffix: /[\n\r\t]+/,
+    // Keys from a barcode scanner are usually processed as quick as possible,
+    // but some scanners can use an intercharacter delay (we support <= 50 ms)
+    max_time_between_keys_in_ms: 55,
+
+    init: function(parent) {
+        mixins.PropertiesMixin.init.call(this);
+        this.setParent(parent);
+        // Keep a reference of the handler functions to use when adding and removing event listeners
+        this.__keydown_handler = _.bind(this.keydown_handler, this);
+        this.__keyup_handler = _.bind(this.keyup_handler, this);
+        this.__handler = _.bind(this.handler, this);
+        // Bind event handler once the DOM is loaded
+        $(this.start.bind(this));
+    },
+
+    handle_buffered_keys: function() {
+        var str = _.reduce(this.buffered_key_events, function(memo, e) { return memo + String.fromCharCode(e.which) }, '');
+        var match = str.match(this.regexp);
+
+        if (match) {
+            var barcode = match[1];
+
+            core.bus.trigger('barcode_scanned', barcode);
+
+            // Dispatch a barcode_scanned DOM event to elements that have barcode_events="true" set.
+            if (this.buffered_key_events[0].target.getAttribute("barcode_events") === "true")
+                $(this.buffered_key_events[0].target).trigger('barcode_scanned', barcode);
+        } else {
+            this.resend_buffered_keys();
+        }
+
+        this.buffered_key_events = [];
+    },
+
+    resend_buffered_keys: function() {
+        function send_keys(list) {
+            if (list.length === 0) return;
+            var old_event = list.shift();
+            var $element = $(old_event.target);
+            var keyCode = old_event.keyCode || old_event.which;
+            setTimeout(function () {
+                setTimeout(function () {
+                    $element.trigger({ type: 'keydown', keyCode: keyCode });
+                    if ((keyCode > 47 && keyCode < 58)   || // number keys
+                        keyCode == 32  || // spacebar
+                        (keyCode > 64 && keyCode < 91)   || // letter keys
+                        (keyCode > 95 && keyCode < 112)  || // numpad keys
+                        (keyCode > 185 && keyCode < 193) || // ;=,-./` (in order)
+                        (keyCode > 218 && keyCode < 223)) {   // [\]' (in order))
+                        document.execCommand('insertText', 0, String.fromCharCode(keyCode));
```

Comment:
Then there's the document.execCommand stuff. I've created a [fiddle](http://jsfiddle.net/nr7jgjmx/1/) that can be used to do cross-browser testing. I've tested in Chromium 43 where it works. In Firefox 38 it does **not** and fails with `NS_ERROR_FAILURE: `. I'm not sure if it can easily be fixed. At least in Firefox it seems to be tied in to [designMode](https://developer.mozilla.org/en-US/docs/Rich-Text_Editing_in_Mozilla), so `document.designMode = true;` seems to expose `document.execCommand()`. And when you do set `designMode` to true it doesn't really do what we want it seems.

I'm not sure if it is easily fixable. We also want to test it on other browsers as well (including mobile variants of browsers).